nomenclature/codelist.py

```diff
@@ -788,10 +788,29 @@ def from_directory(
                     )
                 )
             mapping[code.name] = code
+
         if errors:
             raise ValueError(errors)
         return cls(name=name, mapping=mapping)
 
+    @field_validator("mapping")
+    @classmethod
+    def check_directional_regions(cls, v: dict[str, RegionCode]):
+        missing_regions = []
+        for region in v.values():
+            if region.is_directional:
+                if region.origin not in v:
+                    missing_regions.append(
+                        f"Region '{region.origin}' not defined for '{region.name}'"
+                    )
+                if region.destination not in v:
+                    missing_regions.append(
+                        f"Region '{region.destination}' not defined for '{region.name}'"
+                    )
```

Comment:
Maybe slightly more concise?
```suggestion
                for _region in [region.origin, region.destination]:
                    if _region not in v:
                        missing_regions.append(
                            f"Region '{_region}' not defined for '{region.name}'"
                        )
```

Reply:
I see the argument against duplication but I consciously opted against it for readability. 
To me it's easier to grasp what we're doing when we explicitly have `region.origin` and `region.destination` in the messages.